debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java

```diff
@@ -0,0 +1,540 @@
+        StringBuilder sb = new StringBuilder();
+        byte b = 0;
+        while ((b = buffer.get()) != 0){
+            sb.append((char) b);
+        }
+        return sb.toString();
+    }
+
+    /**
+     * Reads the replication stream where the column stream specifies a length followed by the value.
+     *
+     * @param buffer The replication stream buffer
+     * @return the column value as a string read from the replication stream
+     */
+    private static String readColumnValueAsString(ByteBuffer buffer) {
+        int length = buffer.getInt();
+        byte[] value = new byte[length];
+        buffer.get(value, 0, length);
+        return new String(value);
+    }
+
+    /**
+     * Resolve the replication stream's tuple data to a list of replication message columns.
+     *
+     * @param buffer The replication stream buffer
+     * @param typeRegistry The database type registry
+     * @param table The database table
+     * @return list of replication message columns
+     */
+    private static List<Column> resolveColumnsFromStreamTupleData(ByteBuffer buffer, TypeRegistry typeRegistry, Table table) {
+        // Read number of the columns
+        short numberOfColumns = buffer.getShort();
+
+        List<Column> columns = new ArrayList<>(numberOfColumns);
+        for (short i = 0; i < numberOfColumns; ++i) {
+
+            final io.debezium.relational.Column column = table.columns().get(i);
+            final String columnName = column.name();
+            final String typeName = column.typeName();
+            final PostgresType columnType = typeRegistry.get(typeName);
+            final String typeExpression = column.typeExpression();
+            final boolean optional = column.isOptional();
+
+            // Read the sub-message type
+            // 't' : Value is represented as text
+            // 'u' : An unchanged TOAST-ed value, actual value is not sent.
+            // 'n' : Value is null.
+            char type = (char) buffer.get();
+            if (type == 't') {
+                final String valueStr = readColumnValueAsString(buffer);
+                columns.add(
+                        new AbstractReplicationMessageColumn(columnName, columnType, typeExpression, optional, true) {
+                            @Override
+                            public Object getValue(PgConnectionSupplier connection, boolean includeUnknownDatatypes) {
+                                return PgOutputReplicationMessage.getValue(columnName, columnType, typeExpression, valueStr, connection, includeUnknownDatatypes);
+                            }
+
+                            @Override
+                            public String toString() {
+                                return columnName + "(" + typeExpression + ")=" + valueStr;
+                            }
+                        });
+            }
+            else if (type == 'u') {
+                LOGGER.trace("Skipping TOAST-ed unchanged column '{}'.", columnName);
```

Comment:
Should we really only skip it? Would not be better to provide a special class (used by other decoders too) that would act as proxy for unchanged toast? Similar to null object pattern

Reply:
So locally I have introduced a `ToastedReplicationMessageColumn` that extends the abstract impl.  Inside `ReplicationStreamProducer#columnValues(...)` I actually filter out these toasted columns while we build the returned values array so that the array is consistent with master.  Is this what you'd prefer we do for now or did you have another idea?